### PR TITLE
Builtin translate (2nd attempt)

### LINKED ISSUE #7880: Missing strings for translation
##### Issue type
- Translation issue
 
##### Minetest version
67b20ff15a4786edd76adcbdb921fcc4acac7113

##### Summary
A couple of strings are currently untranslatable or Minetest fails to show the translation:

* `builtin` strings enclosed in `core.gettext` (not collected by `updatepo.sh`)
* Strings in `game.cpp` when `showOverlayMessage` is called (e.g. “Shutting down...”)
* “Unknown node: [node name]” when you hover an unknown node

Please post in this issue if you have found more missing strings

### LINKED ISSUE #7883: Translate builtin
##### Issue type
 - Feature request
 
##### Minetest version
5.0.0-dev

##### Summary
There are still many portions of `builtin` which are untranslated. A large portion is for server-side chat commands and privileges.

Since mod translations are possible now, I believe the server-side portion of `builtin` should be possible to be translated as well. Please give your ideas.

The goal of this issue is to make all user-facing strings in `builtin` translatable.

TODO:
* [ ] Chat commands and output (server side)
* [ ] `help` command and output
* [ ] Chat commands and output (client side)
* [ ] Default node names (air, ignore, unknown)
* [ ] Misc. messages and strings

### PR DESCRIPTION
**Pull request by Wuzzy2**
This PR translates builtin. Fixes #7883. Fixes #7880.

This PR is the second attempt to translate builtin. This PR replaces #9513 which used the wrong translation method. For the new PR I adopted the method suggested by @SmallJoker.

## TODO

This PR works and I also tested stuff. But you probably want to test as well as this PR is quite invasive (pretty much all of builtin is touched, after all).

## How to test

* Add a translation for testing purposes: https://pastebin.com/raw/cbvw1qFz (move to `builtin/locale/__builtin.de.tr`)
* Test in-game (commands, privs, help window, death, etc.)
* Test if all the builtin stuff still works

Report any broken strings, thanks!

## In future

I will submit the German translation in a PR after this is merged.

### CODE REVIEW

File: builtin/game/chat.lua [hunk 1]
```diff
@@ -800,45 +817,41 @@ core.register_on_punchnode(function(pos, node, puncher)
 end)
 
 core.register_chatcommand("rollback_check", {
-	params = "[<range>] [<seconds>] [<limit>]",
-	description = "Check who last touched a node or a node near it"
-			.. " within the time specified by <seconds>. Default: range = 0,"
-			.. " seconds = 86400 = 24h, limit = 5. Set <seconds> to inf for no time limit",
+	params = S("[<range>] [<seconds>] [<limit>]"),
+	description = S("Check who last touched a node or a node near it within the time specified by <seconds>. Default: range = 0, seconds = 86400 = 24h, limit = 5. Set <seconds> to inf for no time limit"),
```
Comment (SmallJoker) [inline builtin/game/chat.lua:821]: Does the gettext parser not support continued lines? I thought that would work.

**Review by SmallJoker — APPROVED** [review]
Looks good. Ran a few commands, and `/help` / `/help privs` to check whether the text is displayed correctly.

**Comment by Wuzzy2** [comment]
Well, the gettext parser sure does, but not the “unofficial hand-made Minetest translation TR file format parser”.

**Comment by Wuzzy2** [comment]
FYI: I was using this script to populate `template.txt`:

https://github.com/minetest-tools/update_translations

**Comment by Wuzzy2** [comment]
Oh, so it turned out the unofficial script DOES support line-breaks.
OK, I have added linebreaks for the uber-long lines in builtin then.

**Comment by Wuzzy2** [comment]
Bump.

File: builtin/client/chatcommands.lua [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
 		return false
 	end
 
 	local first_char = message:sub(1,1)
-	if first_char == "/" or first_char == "." then
-		core.display_chat_message(core.gettext("issued command: ") .. message)
+	if first_char == "." then
+		core.display_chat_message(core.gettext("Issued command: ") .. message)
+	elseif first_char == "/" then
+		core.display_chat_message(S("Issued command: @1", message))
```
Comment (rubenwardy) [inline builtin/client/chatcommands.lua:14]: It's always client-side, in any case

File: builtin/common/chatcommands.lua [hunk 3]
```diff
@@ -1,5 +1,7 @@
 -- Minetest: builtin/common/chatcommands.lua
 
+local S = core.get_translator("__builtin")
```
Comment (rubenwardy) [inline builtin/common/chatcommands.lua:3]: This won't work on the client

File: builtin/common/information_formspecs.lua [hunk 4]
```diff
@@ -20,7 +20,8 @@ local LIST_FORMSPEC_DESCRIPTION = [[
 		button_exit[5,7;3,1;quit;%s]
 	]]
 
-local formspec_escape = core.formspec_escape
+local F = core.formspec_escape
+local S = core.get_translator("__builtin")
```
Comment (rubenwardy) [inline builtin/common/information_formspecs.lua:24]:
```suggestion
local S = core.get_translator("__builtin")
local function FS(...)
	return core.formspec_escape(S(...))
end
```

File: builtin/common/information_formspecs.lua [hunk 5]
```diff
@@ -51,32 +52,33 @@ core.after(0, load_mod_command_tree)
 
 local function build_chatcommands_formspec(name, sel, copy)
 	local rows = {}
-	rows[1] = "#FFF,0,Command,Parameters"
+	rows[1] = "#FFF,0,"..F(S("Command"))..","..F(S("Parameters"))
```
Comment (rubenwardy) [inline builtin/common/information_formspecs.lua:55]: (and all other `F(S())`)

```suggestion
	rows[1] = "#FFF,0,"..FS("Command")..","..FS("Parameters")
```

File: builtin/game/chat.lua [hunk 6]
```diff
@@ -85,9 +87,10 @@ core.register_on_chat_message(function(name, message)
 			core.chat_send_player(name, result)
 		end
 	else
-		core.chat_send_player(name, "You don't have permission"
-				.. " to run this command (missing privileges: "
-				.. table.concat(missing_privs, ", ") .. ")")
+		core.chat_send_player(name,
+				S("You don't have permission to run this command "
```
Comment (rubenwardy) [inline builtin/game/chat.lua:91]: reuse the string please

```suggestion
				S("You don't have permission to run this command "
				.. "(missing privileges: @1)."
```

**Review by rubenwardy — COMMENTED** [review]
Are `params` translated correctly? It feels like there could be issues if you have the same param string but different contexts. For example, if you had `<name>` but it was referring to a player name in one case and something else in another

**Comment by Wuzzy2** [comment]
> Are params translated correctly? It feels like there could be issues if you have the same param string but different contexts.

This context situation does not seem to occur. If there's a case of ambiguity, just change the source string to be more clear. There's not much you can do wrong with the word “name”, I guess.

> It's always client-side so you can remove this

You're correct. Whoops!

> reuse the string please

Oh, there was a runaway parenthesis. I removed it, and checked it in-game.

> FS

I don't like `FS`, to be honest. This requires everyone to update their string-collecting scripts, and we will probably forget about it anyway, and then everyone acts surprise why some strings suddenly don't get collected anymore. Better to leave it as-is.

File: builtin/common/chatcommands.lua (hunk 3 above)
Comment (rubenwardy) [reply builtin/common/chatcommands.lua:3]: This still looks like a problem

**Comment by Wuzzy2** [comment]
> This won't work on the client - this file is used by both. Not sure how to do this right
> This still looks like a problem

So you claim this is broken, but how do you know? What are the steps to reproduce?

**Comment by rubenwardy** [comment]
> So you claim this is broken, but how do you know? What are the steps to reproduce?

The client-side translator is still called on the client, rather than gettext. I looked at `client_lua_api.txt` and it doesn't document `get_translator` as existing in CSM - but it does exist, as it's a pure Lua feature which encodes escape characters. This has the effect of the client displaying some text and using server-provided translations for it. This doesn't actually matter, I guess, and it's better to avoid duplicated translations.

TL;DR: Client builtin still uses server-provided translations, probably fine :shrug:

**Comment by Wuzzy2** [comment]
I'm not sure how to move on …

**Comment by sfan5** [comment]
If in doubt, leave the client stuff untranslated for now.

**Comment by Wuzzy2** [comment]
I've done the rebase, so I hope this helps with testing.

I'd rather do the builtin translate stuff right on the first time, instead of fragmenting the work into multiple PRs, because then we just end up with a half-finished mess in `master`.

I would appreciate if you, @rubenwardy, would tell me steps to actually reproduce and test your concerns. I mean, if you think that clientside is broken in my PR, how can I reliably test whether they are *actually* broken (and not just theoretically)? It might be a valid concern, but the problem is, we don't know if it is actually true. Making decisions based on guessing is a terrible idea.
If it turns out there is no bug after all, then it's great. If it turns out there is a bug, then it must be fixed.

**Comment by rubenwardy** [comment]
It's not broken, it just uses server-provided translations. I think this is fine, and probably the better option compared to duplicating the translations and translation work

**Comment by Wuzzy2** [comment]
Oh, you were talking about the server commands?
Yeah, the translations for server commands also come from the server and NOT the client. This is intentional. And it's the only reasonable behavior IMO. Because the commands are all server-defined (and might have been overwritten), it does not make sense to store the translation for builtin server commands on the clients. If a server overrides builtin for any reason, the clients will still receive the correct translation files. On the other hand, if all builtin translations are only stored on the clients, everything will fail as soon the server changes a single string on its side.

By the way, the textdomain ID is `__builtin`, not `builtin`, keep that in mind when you're creating a dummy translation file.

According to my tests, all of this works, including with client and server, and I still don't get your concerns. :(

I connected to a `builtin_translate3` (=this PR) server (that had a German locale file) with a 5.3.0 client to sure there can't possibly any surprising side-effects. The client still showed the correct translation in German, and when I reconnected with a client with language set to English, it showed the English text instead.

TL;DR: I still don't get any of @rubenwardy's problems, everything seems to work perfectly on my side. \*shrug\*

**Comment by rubenwardy** [comment]
> Because the commands are all server-defined (and might have been overwritten), it does not make sense to store the translation for builtin server commands on the client

False, commands beginning with `.` are defined on the client, not the server

**Comment by Wuzzy2** [comment]
Oh no, you ninja'd me and added the examples later, so my response, which assumes the server-side, doesn't make much sense anymore … xD

Yeah, client-side commands being experimental is a good point. It's still a bit annoying. However, I am still not sure what the correct way to test this thing. If I use the same Minetest for both the server and client, I am worried there will be ugly side-effects which sabotage my testing.
Are you telling me the client-only commands might be on the chopping block? Then it might indeed not be a big deal.

To be clear: Apart from the CLIENT-ONLY commands, what other problems are in this PR?

**Comment by sfan5** [comment]
> I'd rather do the builtin translate stuff right on the first time, instead of fragmenting the work into multiple PRs

What I'm suggesting is to not do translations for client-builtin, at all. With the refocusing on SSCSM it's unclear if it will ever be needed in the future.

**Comment by Wuzzy2** [comment]
Yeah, I agree now the client-command-related stuff isn't suuuper important. It's also the smallest fraction of the strings; the majority is server-side stuff. I'll look into it

**Comment by Wuzzy2** [comment]
Ohhh! I have now found the reason why the client chatcommand strings were not translated for @rubenwardy. In `chatcommands.lua`, the `S` function was used but this function is actually pointless on the client-side.
For ALL client-side stuff, builtin actually needs to use `core.gettext` and not `S`.

The code in `builtin/common/chatcommands.lua` is a bit special since it is both used by client AND server. And when you're on the client-side, tho code still called the `S` but this is actually wrong because the `.tr` file is exclusively meant to be server-sided. The solution is simply to check whether we're a client or a server (`INIT` variable), and use the appropriate translator function. The client should never use `S`.
And that fixes it. It seems this step was just forgotten. The `help` command code already had the `INIT` check present before, just not the other stuff.

So yeah, I decided to properly fix this stuff instead of ignoring it.

Now I think really all builtin strings should be able to be translated properly, no matter if client or serverside.

How to test if clientside strings work:
* Do NOT provide a `.tr` file translation (this is to ensure no surprising side-effects)
* Run `util/updatepo.sh` (since these are Gettext strings now)
* Complete the PO file
* Start Minetest and run all clientside commands like `.help` and related stuff.

File: builtin/game/chat.lua [hunk 7]
```diff
@@ -500,8 +506,8 @@ core.register_chatcommand("teleport", {
 		end
 
 		local has_bring_priv = core.check_player_privs(name, {bring=true})
-		local missing_bring_msg = "You don't have permission to teleport " ..
-			"other players (missing bring privilege)"
+		local missing_bring_msg = S("You don't have permission to teleport " ..
+			"other players (missing privilege: bring).")
```
Comment (SmallJoker) [inline builtin/game/chat.lua:510]: Translators might accidentally translate the privilege name as well. `@1` would be totally save.

File: builtin/common/chatcommands.lua [hunk 8]
```diff
@@ -65,9 +56,22 @@ local function do_help_cmd(name, param)
...
-				.. gettext_replace("Use '$1help <cmd>' to get more information,"
-				.. " or '$1help all' to list everything.", cmd_marker)
+		local msg
+		if INIT == "game" then
+			msg = S("Available commands: @1",
+				table.concat(cmds, " ")) .. "\n"
+				.. S("Use '@1help <cmd>' to get more "
+				.. "information, or '@2help all' to list "
+				.. "everything.",
+				cmd_marker, cmd_marker)
```
Comment (SmallJoker) [inline builtin/common/chatcommands.lua:66]: `cmd_marker` is pointless here if `INIT` chases are handled differently. For consistency, `/help` would fit better here (without replacements).
Also inconsistent use of replacement in `Available commands: [@1]`.

File: builtin/game/chat.lua [hunk 9]
```diff
@@ -664,117 +675,138 @@ local function handle_give_command(cmd, giver, receiver, stackstring)
...
 		p.y = p.y + 1
 		local obj = core.add_entity(p, entityname)
-		local msg = obj and "%q spawned." or "%q failed to spawn."
+		local msg
+		if obj then
+			msg = S("@1 spawned.", entityname)
+		else
+			msg = S("@1 failed to spawn.", entityname)
+		end
 		return true, msg:format(entityname)
```
Comment (SmallJoker) [inline builtin/game/chat.lua:789]: Superfluous `format` call. You could simply return in `if/else` above.

File: builtin/game/chat.lua [hunk 10]
```diff
@@ -837,110 +870,123 @@ core.register_chatcommand("rollback_check", {
...
+					.. "/help rollback_check.")
 			end
+			seconds = tonumber(seconds) or 60
 			target_name = "player:"..player_name
+			rev_msg = S("Reverting actions of player '@1' since @2 seconds.",
+				player_name, seconds)
+		else
+			seconds = tonumber(seconds) or 60
+			rev_msg = S("Reverting actions of @1 since @2 seconds.",
+				target_name, seconds)
```
Comment (SmallJoker) [inline builtin/game/chat.lua:905]: Why is this duplicated code required?

File: builtin/game/chat.lua [hunk 11]
```diff
@@ -1028,71 +1074,70 @@ core.register_chatcommand("clearobjects", {
...
+	description = S("Clear the inventory of yourself or another player"),
 	func = function(name, param)
 		local player
 		if param and param ~= "" and param ~= name then
 			if not core.check_player_privs(name, {server=true}) then
-				return false, "You don't have permission"
-						.. " to clear another player's inventory (missing privilege: server)"
+				return false, S("You don't have permission to "
+					.. "clear another player's inventory "
+					.. "(missing privilege: server).")
```
Comment (SmallJoker) [inline builtin/game/chat.lua:1137]:
```suggestion
					.. "(missing privilege: @1).", "server")
```
for same reasons as mentioned above.

**Comment by Wuzzy2** [comment]
> Why is this duplicated code required?

Because the argument for `string.match` is different.

Yes, the two player-facing strings are similar, but I still like to have them written out in full, because that's much easier for translators to deal with and understand.

File: builtin/common/chatcommands.lua [hunk 12]
```diff
@@ -1,5 +1,7 @@
 -- Minetest: builtin/common/chatcommands.lua
 
+local S = core.get_translator("__builtin")
+
```
Comment (sfan5) [inline builtin/common/chatcommands.lua:4]: Please add a comment here stating that `S` is <b>only</b> to be used in server code (`INIT == "game"`)

File: src/server/mods.h [hunk 13]
```diff
@@ -51,4 +51,5 @@ class ServerModManager : public ModConfiguration
 	 * @param paths result vector
 	 */
 	void getModsMediaPaths(std::vector<std::string> &paths) const;
+	static std::string getBuiltinLuaPath();
```
Comment (sfan5) [inline src/server/mods.h:54]:
```suggestion

	static std::string getBuiltinLuaPath();
```

File: src/server/mods.cpp [hunk 14]
```diff
@@ -98,6 +98,7 @@ void ServerModManager::getModNames(std::vector<std::string> &modlist) const
 
 void ServerModManager::getModsMediaPaths(std::vector<std::string> &paths) const
 {
+	paths.push_back(getBuiltinLuaPath() + DIR_DELIM + "locale");
```
Comment (sfan5) [inline src/server/mods.cpp:101]: Doing this here breaks the unittests: https://github.com/minetest/minetest/blob/07500479191ed927ab661b3758ffcd2fd43158c5/src/unittest/test_servermodmanager.cpp#L158-L164

But regardless of that it'd just be cleaner to do this in [`Server::fillMediaCache()`](https://github.com/minetest/minetest/blob/375bcd65c1903957e3a640cefffcc8df164b78bd/src/server.cpp#L2493)

**Comment by Wuzzy2** [comment]
@SmallJoker @sfan5: Done.

**Review by sfan5 — APPROVED** [review]
Tested.
